src/client/client.cc

```diff
@@ -66,8 +69,26 @@ namespace Pistache::Http::Experimental
                     *https_out = looks_like_https;
             }
 
-            match_string("www", cursor);
-            match_literal('.', cursor);
+            // Skipping the subdomain ("www.") is a good idea if we want to
+            // resolve to an IP address for a TCP or SSL connection. It is a
+            // bad idea if we are constructing the REST request - it may result
+            // in a 301 HTTP error ("document moved" aka a redirection). E.g.
+            // curl "https://google.com" - gets a 301 reponse, while curl
+            // "https://www.google.com" - gets a 200 response i.e. success.
+            //
+            // ALSO - this is a weak way to remove the subdomain. There are
+            // many other possible subdomains than just "www". To actually
+            // identify the subdomain, you need to identify the TLD (".com" or
+            // whatever), work back to the domain, and then isolate the
+            // subdomain part of the URL. See https://publicsuffix.org and e.g.
+            // https://stackoverflow.com/questions/288810/
+            //                                     get-the-subdomain-from-a-url
+            // TODO !!!!
+            if (remove_subdomain)
+            {
+                match_string("www", cursor);
+                match_literal('.', cursor);
+            }
```

Comment:
I'm not sure I understand the reason behind this change. Also, why not fixing the underlying issue directly? Maybe in a different PR, so that it can me reviewed and merged independently from this big one :)

Also, it's API-breaking. Maybe you could add the new boolean as the last parameter, with a default value of `true`, as not to change behaviour.

Reply:
The reason for the parameter is that we DO want to remove the "www" when making a transport-level (TCP or SSL) connection address, but DO NOT want to remove the "www" when constructing a a REST request, to avoid getting a 302 back (as per the comment). I'm not sure there is an "underlying issue" beyond that. As you saw, previously it would always remove the "www" - but that was then resulting in getting 302s back from the Google endpoint at the HTTP level.

I admit this is one change that could have been in a separate PR :-)

Regarding API breaking - splitURL is internal to client.cc - so it shouldn't be API breaking, should it? Or am I missing the point @Tachi107 ?